EAR-1001-create-folder

MattElcock opened the pull request:
### What is the context of this PR?

This PR introduces the ability to add a new folder via the Author UI. I have also included some refactoring; changing old class components to function components, making some existing components more reusable, and updating Storybook with new components.

### How to review

Try adding a new folder and see if you can update it's QCode. You should not be able to add a page within the folder yet, but it should initialise with an empty page. You should not be able to delete, move or duplicate the folder and the buttons should be disabled.

- Import the **Capability examples** questionnaire from pre-prod Author into your local environment and:
  - ensure it can be opened in Author;
  - then, ensure it can be viewed in Runner by pressing the **view survey** button
- Does this require a migration? We need one if existing JSON schema properties change

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process

tjamesmac commented on eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/index.js:
```diff
@@ -48,146 +62,160 @@ const AccordionGroupToggle = styled(Button).attrs({
+  const buildNavigationList = ({ sections }) => {
+    const components = sections.map(
+      ({ id: sectionId, title, folders, validationErrorInfo }) => {
+        const allPagesInSection = flatMap(folders.map(({ pages }) => pages));
```
```suggestion
        const allPagesInSection = folders.flatMap(({ pages }) => pages);
```
cut out the middle person

tjamesmac commented on eq-author/src/App/QuestionnairesPage/QuestionnairesView/QuestionnairesTable/Row/index.js:
```diff
@@ -259,6 +260,7 @@ export const Row = ({
               <DuplicateQuestionnaireButton
                 data-test="btn-duplicate-questionnaire"
                 onClick={handleDuplicateQuestionnaire}
+                hideText
```
In the questionnaire list

MattElcock replied:
This is actually because I forgot to add the tooltip back in when I was fiddling around. Oops! I've fixed this.

tjamesmac commented on eq-author/src/App/QuestionnaireDesignPage/index.js:
```diff
@@ -54,32 +66,77 @@ const MainNav = styled.div`
+
+    const pages = questionnaire.sections.flatMap(({ folders }) =>
+      folders.flatMap(folder => folder.pages)
+    );
```
use `getPages()` here?

tjamesmac commented on eq-author/src/App/QuestionnaireDesignPage/index.js:
```diff
@@ -54,32 +66,77 @@ const MainNav = styled.div`
+      folders.flatMap(folder => folder.pages)
+    );
+
+    const page = find(pages, { id: pageId });
+
```
Or even `getPageById()`?

MattElcock commented:
From Sam:

> Hi Matt. I’ve done a bit of testing on the folders branch, and I noticed a few things about it. Some of these might just have been changes to the design though, but I thought it best to send all of them in case they aren’t.
>
> When there are no other questions and you delete a question inside of a folder (which would also delete the folder), the folder does not get deleted and the question shows inside the folder. This may be design choice and not an error though, as when EAR 1002 is added, it will be easier to delete the folder, but I made a note of it just in case it wasn’t intended as it’s the only time when deleting the question doesn’t delete the folder. To delete a folder when this happens at the moment, a new question has to be created outside of the folder, and then the question inside of the folder has to be deleted after
>
>When all of the errors are removed from a question, the red badge with 1 is still displayed on the left nav bar. Refreshing the page removes the badge correctly, but before refreshing, the error badge continues to be shown.  While the incorrect error is being displayed, it is not possible to create a new question, section etc. Shane had a very similar issue before, and it was because he missed out a field from some of his GraphQL code
>
>The text on the sign in page gets placed onto two different lines, instead of being on the one line it was on before
Folders do not appear as sections when respondents are completing the questionnaire; respondents only see the questions contains within folders if they meet the conditions you have applied. I might be wrong but I think the word contains should be changed to contained
>
>The space between the first question and a section title in the left nav bar has been reduced. When the first question is selected and highlighted in orange, the line that separates a section and a question is covered by the highlight and mostly isn’t visible. This may be a design choice.
>
>Hovering over a question, section etc in the left nav bar does not highlight it in a slightly different colour like it used to. This may be a design choice

tjamesmac reviewed (dismissed):
Front-end:

- Currently the intro/section/folder tab has an outline as well as the orange background colour
![Screenshot 2021-02-16 at 11 32 20](https://user-images.githubusercontent.com/22731314/108057497-a438aa80-704a-11eb-93dd-81c5a4284dd2.png)
- Numbered error dots are a different shape now


| Before  | After |
| ------------- | ------------- |
| ![Screenshot 2021-02-16 at 11 34 51](https://user-images.githubusercontent.com/22731314/108057844-14dfc700-704b-11eb-82da-1aa4bb387c7c.png)  | ![Screenshot 2021-02-16 at 11 33 55](https://user-images.githubusercontent.com/22731314/108057869-1dd09880-704b-11eb-8645-f85533dd397c.png)  |

- Gap between sections is bigger now

Need to do a bit more front-end testing though

MattElcock commented on eq-author/src/App/folder/Design/index.test.js:
```diff
@@ -0,0 +1,125 @@
+
+const firstFolder = mockQuestionnaire.sections[0].folders[0];
+
+const mockUser = {
+  id: "123",
+  displayName: "Raymond Holt",
+  email: "RaymondHolt@TheNineNine.com",
+  picture: "http://img.com/avatar.jpg",
+  admin: true,
+};
```
Hahaha

MattElcock commented on eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/index.test.js:
```diff
@@ -1,190 +1,204 @@
-  buildSections,
-} from "tests/utils/createMockQuestionnaire";
 
-describe("NavigationSidebar", () => {
```
A lot of these tests are already covered by other ones:

-  `AddMenu.test.js` includes tests for adding a page, section, confirmation, calculated summary and when each should be disabled. 
- An introduction page being featured on the questionnaire isn't a behavior dictated by the side nav panel, so that one isn't needed here. 
- The navigation sidebar doesn't have reducers anymore, so these aren't needed.
- Opening and closing the collapsible toggles is covered in `CollapsibleNavigationItem.test.js`

I can add a test for the Open/Close all button, 'though. I think that's the only one that isn't covered elsewhere. I'll check that before hand 'though to be sure.

tjamesmac commented:
Thank you for adding the storybook backgrounds in. Is it possible to remove the links for the `collapsibleNavItems` because if you click on them it reloads the page

MattElcock commented:
> Numbered error dots are a different shape now

Fixed.

MattElcock commented:
> Currently the intro/section/folder tab has an outline as well as the orange background colour

I'm not entirely sure why, but when you click on a folder it reloads the page and keeps your focus on that button. This doesn't happen when you're on a page. Clicking away from the button takes the focus state off.

MattElcock commented:
> When there are no other questions and you delete a question inside of a folder (which would also delete the folder), the folder does not get deleted and the question shows inside the folder. This may be design choice and not an error though, as when EAR 1002 is added, it will be easier to delete the folder, but I made a note of it just in case it wasn’t intended as it’s the only time when deleting the question doesn’t delete the folder. To delete a folder when this happens at the moment, a new question has to be created outside of the folder, and then the question inside of the folder has to be deleted after

This will be done as part of deleting a folder ticket.

MattElcock commented:
> The text on the sign in page gets placed onto two different lines, instead of being on the one line it was on before
Folders do not appear as sections when respondents are completing the questionnaire; respondents only see the questions contains within folders if they meet the conditions you have applied. I might be wrong but I think the word contains should be changed to contained

Fixed

MattElcock commented:
> The space between the first question and a section title in the left nav bar has been reduced. When the first question is selected and highlighted in orange, the line that separates a section and a question is covered by the highlight and mostly isn’t visible. This may be a design choice.

Fixed

MattElcock commented:
> Hovering over a question, section etc in the left nav bar does not highlight it in a slightly different colour like it used to. This may be a design choice

Fixed

farres1 commented on eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/index.js:
```diff
@@ -203,29 +292,35 @@ export const UnwrappedNavigationSidebar = (props) => {
-            {label ? "Close all" : "Open all"}
-          </AccordionGroupToggle>
+          <OpenAllSectionsBtn onClick={() => toggleSections(!openSections)}>
+            {`${openSections ? "Open" : "Close"} all sections`}
```
These should be the other way around
```suggestion
            {`${openSections ? "Close" : "Open"} all sections`}
```

MattElcock commented:
> Is it possible to remove the links for the collapsibleNavItems because if you click on them it reloads the page

Fixed this

jchapman68 commented:
Errors not clearing for existing surveys.
- switched to master
- created questionnaire
- first question with one number answer and no errors
- second question with one number answer and no errors
- selected first question and added a routing rule but not complete creating two errors
- selected home
- switched to EAR-1001-create-folder thread
- opened questionnaire
- selected second question
- added a folder
- updated new question under folder with question text and added a number answer
- one error showing on new question but nothing to fix
- removed rule on first question which should have cleared the errors but they remain
- three errors showing but nothing to fix
- delete the first question but it isn't removed
- refresh page 
- have repeated above three times :
- twice the first question is removed and the error on the folder question disappears
- once the front end crashed with "TypeError: Cannot read property 'pageType' of null"

MattElcock commented on eq-author/src/App/QuestionnaireDesignPage/NavigationSidebar/index.js:
```diff
@@ -203,29 +298,36 @@ export const UnwrappedNavigationSidebar = (props) => {
+                    )}
+                    icon={PageIcon}
+                    history={history}
+                    style={{ marginLeft: "1.5em" }}
```
This doesn't need to be an inline style as it's a styled component.

martyncolmer commented:
Sections nav item no longer has an error count for section errors.
EG when section navigation is enabled and no title entered.

martyncolmer commented:
> Errors not clearing for existing surveys.
> 
> * switched to master
> * created questionnaire

Had look at the issue Josh raised, when you update an answer the error count is not updating. The cache does updater with the new details but the menu refresh isn't firing.

farres1 reviewed (commented):
Duplicating any question that doesn’t have any errors causes the question not to be shown in the left nav bar. You have to refresh the page for the question to be displayed there
This might be linked to the error dot not being removed

MattElcock commented:
> Sections nav item no longer has an error count for section errors.
EG when section navigation is enabled and no title entered.

Fixed

MattElcock commented:
> Errors not clearing for existing surveys.
> 
> * switched to master
> * created questionnaire
> * first question with one number answer and no errors
> * second question with one number answer and no errors
> * selected first question and added a routing rule but not complete creating two errors
> * selected home
> * switched to EAR-1001-create-folder
> * opened questionnaire
> * selected second question
> * added a folder
> * updated new question under folder with question text and added a number answer
> * one error showing on new question but nothing to fix
> * removed rule on first question which should have cleared the errors but they remain
> * three errors showing but nothing to fix
> * delete the first question but it isn't removed
> * refresh page
> * have repeated above three times :
> * twice the first question is removed and the error on the folder question disappears
> * once the front end crashed with "TypeError: Cannot read property 'pageType' of null"

Fixed; Cheers Martyn!

MattElcock commented:
> Duplicating any question that doesn’t have any errors causes the question not to be shown in the left nav bar. You have to refresh the page for the question to be displayed there
This might be linked to the error dot not being removed

Fixed!

MattElcock commented:
> ![image](https://user-images.githubusercontent.com/15129656/108211112-3cef2900-7124-11eb-83e3-f01ee1ec2063.png)
> 
> From Josh: Nav now doesn't truncate

Fixed

jasehumphr commented:
Not sure it was like this before <img width="827" alt="Screenshot 2021-02-17 at 15 45 17" src="https://user-images.githubusercontent.com/55096559/108229114-69607080-7137-11eb-819e-54cb43825ba0.png">

jchapman68 commented:
Survey - “Or” Mutual exclusive questions
1. Question titles in nav are centred and line wrapped
2. Error remaining in nav on new folder question after updating question and adding answer and label, needs refresh to clear
3. Routing errors not clearing on tab and nav unless change tab or refresh page
4. After a refresh logic errors do clear on tab but still need refresh to clear on nav

jchapman68 commented:
When moving question the nav is not updated without a refresh.  If making a further move before refreshing the move list is as per the nav, not the updated list.

farres1 reviewed (commented):
Duplicating a section causes the new duplicated section not to be displayed on the left nav bar
Refresh required to show the new section in the nav bar

tjamesmac commented:
+1 to what @farres1 said

tjamesmac commented:
Gap between folder and child question is too big (it's the margin on the child question)

farres1 reviewed (commented):
Creating a radio or checkbox answer causes an error to be shown in the nav bar after all errors on the question are fixed
The error stays until refresh

MattElcock commented:
> Not sure it was like this before <img alt="Screenshot 2021-02-17 at 15 45 17" width="827" src="https://user-images.githubusercontent.com/55096559/108229114-69607080-7137-11eb-819e-54cb43825ba0.png">
> 
> This occurs when the page is made a bit smaller. Can anyone else replicate? Looks like this is not a fault on this ticket. I bet this is somewhere in the depths of the back log

I can replicate this, both locally and on staging. You're right, it's not a fault on this ticket. There might be a ticket for it already; if you can't find one, could you make one please?

MattElcock commented:
> Survey - “Or” Mutual exclusive questions
> 
> 1. Question titles in nav are centred and line wrapped

This has been fixed as part of your earlier comment. I can't replicate it. You may need to pull the latest changes if you haven't already. If you can still make it happen, could you expand on how you're making it happen and give some screenshots please?

MattElcock commented:
> When moving question the nav is not updated without a refresh. If making a further move before refreshing the move list is as per the nav, not the updated list.

Fixed.

martyncolmer commented:
Adding a confirmation question overwrites the question it's part off

martyncolmer commented:
Adding a Calculated summary doesn't appear without a refresh.

martyncolmer commented:
When there is a calc summary on the page, add a question now breaks without a refresh.

MattElcock commented:
> Gap between folder and child question is too big (it's the margin on the child question)

Fixed

jchapman68 reviewed (changes requested):
Front end crashes with `TypeError: Cannot read property 'id' of null`

- Switch to master
- Import ['Or' - Mutual exclusive questions (live)](https://author.ons.gov.uk/#/q/d533b505-af21-46a4-a3fd-d35f57073869/introduction/480d1c8b-19db-4bb7-bc3c-f0d4051e4040/design)
- Switch back to this branch
- Open survey
- Add folder after last question
- Select last question again before folder
- Add Confirmation question
- With Confirmation question selected add another folder

MattElcock commented:
> Creating a radio or checkbox answer causes an error to be shown in the nav bar after all errors on the question are fixed
> The error stays until refresh

Fixed

MattElcock commented:
> Duplicating a section causes the new duplicated section not to be displayed on the left nav bar
> Refresh required to show the new section in the nav bar

Fixed

MattElcock commented:
> Adding a confirmation question overwrites the question it's part off

Fixed

MattElcock commented:
> Adding a Calculated summary doesn't appear without a refresh.

Fixed

MattElcock commented:
> Front end crashes with `TypeError: Cannot read property 'id' of null`
> 
> * Switch to master
> * Import ['Or' - Mutual exclusive questions (live)](https://author.ons.gov.uk/#/q/d533b505-af21-46a4-a3fd-d35f57073869/introduction/480d1c8b-19db-4bb7-bc3c-f0d4051e4040/design)
> * Switch back to this branch
> * Open survey
> * Add folder after last question
> * Select last question again before folder
> * Add Confirmation question
> * With Confirmation question selected add another folder

Fixed

farres1 reviewed (commented):
The space between the section and the first question has been reduced. When the first question is highlighted it covers the line between section and question

Long pressing an unselected question displays a box around it that is not displayed on previous designs

farres1 reviewed (commented):
The section duplication error is affecting sections that have questions with no errors

tbaggaley commented on eq-author/src/App/folder/Design/index.js:
```diff
@@ -0,0 +1,129 @@
+          </p>
+          <p>
+            Folders do not appear as sections when respondents are completing
+            the questionnaire; respondents only see the questions contains
```
```suggestion
            the questionnaire; respondents only see the questions contained
```

jchapman68 reviewed (changes requested):
Front end crashes on existing survey
- Switch to master
- Import existing survey
- Switch to folders branch
- Open survey
- Select last question
- Add folder
- Reselect last question (not new folder question)
- Add confirmation page
- Whilst on confirmation page add a folder

tjamesmac commented:
> Front end crashes on existing survey
> 
> * Switch to master
> * Import existing survey
> * Switch to folders branch
> * Open survey
> * Select last question
> * Add folder
> * Reselect last question (not new folder question)
> * Add confirmation page
> * Whilst on confirmation page add a folder

Think this is just if you try to add a folder whilst on a confirmation question

MattElcock commented:
> > Front end crashes on existing survey
> > 
> > * Switch to master
> > * Import existing survey
> > * Switch to folders branch
> > * Open survey
> > * Select last question
> > * Add folder
> > * Reselect last question (not new folder question)
> > * Add confirmation page
> > * Whilst on confirmation page add a folder
> 
> Think this is just if you try to add a folder whilst on a confirmation question

jchapman68 reviewed (dismissed):
New calculated summary page not appearing in nav
- Import existing survey on master
- Switch to folders branch
- Open survey and select a question
- Add a calculated summary, page opens but not showing on nav

Calc summary not showing also crashes front end
- on imported survey select last question
- add a question
- add a calculated summary (which doesn't show in nav)
- select question just added and delete
- front end crashes saying can't find calc summary, but not if refreshed and calc summary showing in nav

MattElcock commented:
> The section duplication error is affecting sections that have questions with no errors

Looks like this is fixed as I can't replicate it.

MattElcock commented:
> When there is a calc summary on the page, add a question now breaks without a refresh.

Fixed

MattElcock commented:
> The space between the section and the first question has been reduced. When the first question is highlighted it covers the line between section and question
> 
> Long pressing an unselected question displays a box around it that is not displayed on previous designs

Fixed the spacing. The long press is it's on-focus styles. The action of clicking down on it triggers the on focus styles, but the onClick events fires on the mouse keys 'on up' event. That's just normal behaviour though. (You can replicate this on other sites, like GitHub, by doing the same on their buttons; the onFocus styles trigger if you long press but the button's onClick doesn't).

MattElcock commented:
> The section duplication error is affecting sections that have questions with no errors

Looks like this was fixed for free with something else.

MattElcock commented:
> New calculated summary page not appearing in nav

Fixed

martyncolmer commented:
We are missing the error count on the confirmation page

tjamesmac reviewed (approved):
It's aliiiiiive!

jasehumphr reviewed (approved):
I loves a little bit of red

tjamesmac commented:
Here lies the greatest PR to ever live

jasehumphr commented:
Was this a test to see if I actually read my email?

From: Thomas McAuliffe <notifications@github.com>
Sent: 19 February 2021 13:35
To: ONSdigital/eq-author-app <eq-author-app@noreply.github.com>
Cc: Humphries, Jason <Jason.Humphries@ext.ons.gov.uk>; Comment <comment@noreply.github.com>
Subject: Re: [ONSdigital/eq-author-app] EAR-1001-create-folder (#1748)


Here lies the greatest PR to ever live

-
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FONSdigital%2Feq-author-app%2Fpull%2F1748%23issuecomment-782078942&data=04%7C01%7CJason.Humphries%40ext.ons.gov.uk%7C747c5a4f5ce842f6087608d8d4db334b%7C078807bfce824688bce00d811684dc46%7C0%7C0%7C637493385223234136%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=xPxFbrzYpD9ZWCTKC%2BU0qLQWZfCXKs7ZwxM%2FdrAVzZQ%3D&reserved=0>, or unsubscribe<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FANELJ33VBHOYVHQITRMKOMDS7ZSJJANCNFSM4WVN72PA&data=04%7C01%7CJason.Humphries%40ext.ons.gov.uk%7C747c5a4f5ce842f6087608d8d4db334b%7C078807bfce824688bce00d811684dc46%7C0%7C0%7C637493385223234136%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=w1tfYbshIsJwW6bPhOeBc%2FaSKM6ux5jGV0Ra4qaSVyo%3D&reserved=0>.
